File: packages/expo-ui/ios/TextInput/TextInputMask/ForkInputMask/Source/InputMask/InputMask/Classes/RTLMask.swift
//
// Project «InputMask»
// Created by Jeorge Taflanidi
//

#if canImport(Foundation)

import Foundation


/**
 ### RTLMask

 A right-to-left ``Mask`` subclass. Applies format from the string end.
 */
public class RTLMask: Mask {
    private static var cache: [String: RTLMask] = [:]
    
    public required init(format: String, customNotations: [Notation] = []) throws {
        try super.init(format: format.reversedFormat(), customNotations: customNotations)
    }
    
    public override class func getOrCreate(withFormat format: String, customNotations: [Notation] = []) throws -> RTLMask {
        if let cachedMask: RTLMask = cache[format.reversedFormat()] {
            return cachedMask
        } else {
            let mask: RTLMask = try RTLMask(format: format, customNotations: customNotations)
            cache[format.reversedFormat()] = mask
            return mask
        }
    }

    public override func apply(toText text: CaretString) -> Result {
        return super.apply(toText: text.reversed()).reversed()
    }
    
    override func makeIterator(forText text: CaretString) -> CaretStringIterator {
        return RTLCaretStringIterator(caretString: text)
    }
}

#endif

File: packages/expo-ui/android/src/main/java/expo/modules/ui/textfield/helper/RTLMask.kt
package expo.modules.ui.textfield.helper

import expo.modules.ui.textfield.helper.Mask
import expo.modules.ui.textfield.model.CaretString
import expo.modules.ui.textfield.model.Notation
import java.util.HashMap

/**
 * ### RTLMask
 *
 * A right-to-left ``Mask`` subclass. Applies format from the string end.
 */
class RTLMask(format: String, customNotations: List<Notation>) : Mask(reversedFormat(format), customNotations) {
    companion object Factory {
        private val cache: MutableMap<String, RTLMask> = HashMap()

        fun getOrCreate(format: String, customNotations: List<Notation>): RTLMask {
            var cachedMask: RTLMask? = cache[reversedFormat(format)]
            if (null == cachedMask) {
                cachedMask = RTLMask(format, customNotations)
                cache[reversedFormat(format)] = cachedMask
            }
            return cachedMask
        }
    }

    override fun apply(text: CaretString): Result {
        return super.apply(text.reversed()).reversed()
    }

    override fun makeIterator(text: CaretString): CaretStringIterator {
        return RTLCaretStringIterator(text)
    }
}

private fun reversedFormat(format: String) =
    format
        .reversed()
        .replace("[\\", "\\]")
        .replace("]\\", "\\[")
        .replace("{\\", "\\}")
        .replace("}\\", "\\{")
        .map {
            when (it) {
                '[' -> ']'
                ']' -> '['
                '{' -> '}'
                '}' -> '{'
                else -> it
            }
        }.joinToString("")
